File: edu/cmu/hcii/puc/uigen/TabPanelNode.java
/**
 * TabPanelNode.java
 *
 * This node represents nodes that represent a single panel.
 *
 * @author Jeffrey Nichols
 */

// Package Definition

package edu.cmu.hcii.puc.uigen;


// Import Declarations

import java.awt.Component;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.Rectangle;
import java.awt.Panel;

import java.util.Enumeration;
import java.util.Vector;

import edu.cmu.hcii.puc.cio.ContainerCIO;
import edu.cmu.hcii.puc.cio.TabbedLinkedCIO;


// Class Definition

public class TabPanelNode extends MultiplePanelNode {

    //**************************
    // Member Variables
    //**************************

    ContainerCIO  m_pPanel;


    //**************************
    // Constructor
    //**************************

    public TabPanelNode( TabbedLinkedCIO.TabbedPanelCIO pCIO ) {
	super();

	m_pPanel = pCIO;
	addPanel( new PanelNode() );
    }

    //**************************
    // Member Methods
    //**************************

    public InterfaceNode getChildNode() {

	return (InterfaceNode)m_vPanels.elementAt( 0 );
    }

    public ContainerCIO getContainerCIO() {

	return m_pPanel;
    }


    //**************************
    // Interface Methods
    //**************************

    public void addComponents( ContainerCIO pContainer ) {

	((Container)m_pPanel.getWidget()).setLayout( null );

	((InterfaceNode)m_vPanels.elementAt( 0 )).addComponents( m_pPanel );
    }

    public Dimension getPreferredSize( LayoutVariables pVars ) {

	return ((InterfaceNode)m_vPanels.elementAt( 0 )).getPreferredSize( pVars );
    }

    public void doSizing( LayoutVariables pVars ) {
	
	Dimension d = m_pPanel.getWidget().getSize();
	this.setSize( d.width, d.height );

	InterfaceNode pNode = (InterfaceNode)m_vPanels.elementAt( 0 );

	pNode.setSize( m_rBounds.width, m_rBounds.height );
	pNode.setLocation( 0, 0 );

	pNode.doSizing( pVars );
    }

    public void doLayout( ContainerCIO pContainer, LayoutVariables pVars ) {

	((InterfaceNode)m_vPanels.elementAt( 0 )).doLayout( m_pPanel, pVars );
    }


    public String toString() {

	return "TabPanelNode - " + super.toString();
    }
}

File: csharp/UIGeneration/TabPanelNode.cs
using System;
using System.Collections;
using System.Windows.Forms;

using PUC;
using PUC.CIO;
using PUC.Layout;

namespace PUC.UIGeneration
{
	/// <summary>
	/// This represents the node of an interface that is a panel
	/// within a tab control.
	/// </summary>
	public class TabPanelNode : MultiplePanelNode
	{
		/*
		 * Member Variables
		 */

		protected ContainerCIO _panel;


		/*
		 * Constructor
		 */

		public TabPanelNode( TabbedPanelCIO cio, GroupNode g )
		{
			_panel = cio;
			AddPanel( new PanelNode( g ) );
		}


		/*
		 * Member Methods
		 */

		public InterfaceNode GetChildNode()
		{
			return (InterfaceNode)_panels[ 0 ];
		}

		public ContainerCIO GetContainerCIO()
		{
			return _panel;
		}


		/*
		 * InterfaceNode Methods
		 */

		public override void AddComponents( ContainerCIO container, LayoutVariables vars )
		{
			Container = container;
			((InterfaceNode)_panels[ 0 ]).AddComponents( _panel, vars );

			this.CalculateMinimumSize( vars );
			this.CalculatePreferredSize( vars );
		}

		public override void CalculateMinimumSize(LayoutVariables vars)
		{
			((InterfaceNode)_panels[ 0 ]).CalculateMinimumSize( vars );
			_minSize = ((InterfaceNode)_panels[ 0 ]).MinimumSize;
		}

		public override void CalculatePreferredSize(LayoutVariables vars)
		{
			((InterfaceNode)_panels[ 0 ]).CalculatePreferredSize( vars );
			_prefSize = ((InterfaceNode)_panels[ 0 ]).PreferredSize; 
		}

		public override void DoLayout(LayoutVariables vars)
		{
			InterfaceNode node = (InterfaceNode)_panels[ 0 ];

			// set bounds to the control bounds, which are automattically set by
			// the TabControl
			Control c = _panel.GetControl();
			node.SetSize( c.Bounds.Width, c.Bounds.Height );
			node.SetLocation( 0, 0 );

			((InterfaceNode)_panels[ 0 ]).DoLayout( vars );
		}

		public override string ToString()
		{
			return "TabPanelNode - " + base.ToString();
		}
	}
}
